Refactor BlockDefinition to support blocks defined in JavaScript.

AnmAtAnm opened the pull request:
Adding .javascript_ field to support block definitions defined in
JavaScript. Adding a format string parameter (either 'JSON' or
'JS) to many of the methods.

Next step:
Updates to JSON definition are still not updating the preview.
Check BlockEditorController.updatePreview() or .refreshPreview().

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-devtools/284)
<!-- Reviewable:end -->

AnmAtAnm committed:
Refactor BlockDefinition to support blocks defined in JavaScript.

Adding .javascript_ field to support block definitions defined in
JavaScript. Adding a format string parameter (either 'JSON' or
'JS) to many of the methods.

Next step:
Updates to JSON definition are still not updating the preview.
Check BlockEditorController.updatePreview() or .refreshPreview().

rachel-fenichel commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/model/block_definition.js, line 57 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/284#-L30teh_Kf8oWpItTMpX:-L30teh_Kf8oWpItTMpY:bya1blr) ([raw file](https://github.com/google/blockly-devtools/blob/1acd2879eb2aa712cd9edb3233ab23af9e8c8b99/src/model/block_definition.js#L57)):*
> ```JavaScript
> 
>     /**
>      * The JSON representation of the block.
> ```

Fix comment.

---

*[src/model/block_definition.js, line 121 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/284#-L30u69fCw5CQJz_Msrd:-L30u69fCw5CQJz_Msre:b-62pmi) ([raw file](https://github.com/google/blockly-devtools/blob/1acd2879eb2aa712cd9edb3233ab23af9e8c8b99/src/model/block_definition.js#L121)):*
> ```JavaScript
>    */
>   defineFromJson_() {
>     const typeName = this.type();
> ```

You removed the check that the name exists/is valid.  Why?

---

*[src/model/block_definition.js, line 154 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/284#-L30uTyIUbW6kFa64jot:-L30uTyIUbW6kFa64jou:b-sk2kdv) ([raw file](https://github.com/google/blockly-devtools/blob/1acd2879eb2aa712cd9edb3233ab23af9e8c8b99/src/model/block_definition.js#L154)):*
> ```JavaScript
>       console.error('Error while evaluating JavaScript formatted block definition', e);
>       undefine();  // Attempt to reset state.
>       throw new Error('Failed to define block type from JavaScript.');
> ```

Why do you have both a console.error and a new Error?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/284)*
<!-- Sent from Reviewable.io -->

picklesrus commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/model/block_definition.js, line 34 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/284#-L30zt88vprcgUqgSof3:-L30zt89iXLGhIzE9BrM:b-eygeh4) ([raw file](https://github.com/google/blockly-devtools/blob/1acd2879eb2aa712cd9edb3233ab23af9e8c8b99/src/model/block_definition.js#L34)):*
> ```JavaScript
>    * BlockDefinition Class.
>    * @constructor
>    * @param {string} blockTypeName The type name of the definition.
> ```

What is a TypeName in this context? I thought it was going to be json or js but that's in format.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/284)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/model/block_definition.js, line 34 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/284#-L30zt88vprcgUqgSof3:-L34VUJ92YQux4WNoAYI:b-kjsn3) ([raw file](https://github.com/google/blockly-devtools/blob/f15c938a93d567090049e6ceb5e90d172c08fb15/src/model/block_definition.js#L34)):*
<details><summary><i>Previously, picklesrus wrote…</i></summary><blockquote>

What is a TypeName in this context? I thought it was going to be json or js but that's in format.
</blockquote></details>

This is the name of the block type, the one used as a key to Blockly.Blocks[] or the 'name' field of JSON.

---

*[src/model/block_definition.js, line 57 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/284#-L30teh_Kf8oWpItTMpX:-L34VmNaNYihIBFAzSu7:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/f15c938a93d567090049e6ceb5e90d172c08fb15/src/model/block_definition.js#L57)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Fix comment.
</blockquote></details>

Done.

---

*[src/model/block_definition.js, line 121 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/284#-L30u69fCw5CQJz_Msrd:-L34VK1hBFcUVojugCMP:blu7seg) ([raw file](https://github.com/google/blockly-devtools/blob/f15c938a93d567090049e6ceb5e90d172c08fb15/src/model/block_definition.js#L121)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

You removed the check that the name exists/is valid.  Why?
</blockquote></details>

Added back.

---

*[src/model/block_definition.js, line 154 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/284#-L30uTyIUbW6kFa64jot:-L34TXxBiX9Jk7a3g8h8:b-r48dup) ([raw file](https://github.com/google/blockly-devtools/blob/f15c938a93d567090049e6ceb5e90d172c08fb15/src/model/block_definition.js#L154)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Why do you have both a console.error and a new Error?
</blockquote></details>

If I make make a guess about my former self...
There is no 'cause' field in JavaScript errors, so logging the original error to console.error(..) captures the deeper stack trace. Throwing a different error lets me give a description that may make more sense in the calling context.

Added a comment above console.error(..).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/284)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/284#-:-L34aR0m31EAoNXuJ_50:bnfp4nl)*
<!-- Sent from Reviewable.io -->
